File: rust/tests/pressure/bmp180_test/src/main.rs
use linux_embedded_hal::I2cdev;
use periph::chips::pressure::{Bmp180Minimal, Bmp180Full, OSS_ULP};

macro_rules! check_true {
    ($cond:expr, $label:expr, $passed:expr, $failed:expr) => {
        if $cond { println!("PASS {}", $label); $passed += 1; }
        else      { println!("FAIL {}", $label); $failed += 1; }
    };
}

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR")
        .ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x77);

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut bmp = Bmp180Minimal::new(dev, addr).expect("init BMP180");

    bmp.inner.b5 = 0;
    bmp.inner.ac1 = 408;
    bmp.inner.ac2 = -72;
    bmp.inner.ac3 = -14383;
    bmp.inner.ac4 = 32741;
    bmp.inner.ac5 = 32757;
    bmp.inner.ac6 = 23153;
    bmp.inner.b1 = 6190;
    bmp.inner.b2 = 4;
    bmp.inner.mc = -8711;
    bmp.inner.md = 2868;

    let b5 = periph::chips::pressure::bmp180::compensate_temp(
        27898, bmp.inner.ac1, bmp.inner.ac2, bmp.inner.ac3,
        bmp.inner.ac5, bmp.inner.ac6, bmp.inner.mc, bmp.inner.md
    );
    check_true!(b5 != 0, "temp_compensation_b5", passed, failed);

    drop(bmp);

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut bmp_full = Bmp180Full::new(dev, addr, OSS_ULP).expect("init BMP180 Full");
    check_true!(bmp_full.oversampling() == 0, "default_oss", passed, failed);
    bmp_full.set_oversampling(2);
    check_true!(bmp_full.oversampling() == 2, "set_oss", passed, failed);

    let alt = bmp_full.altitude(1013.25).unwrap_or(-1.0);
    check_true!(alt >= 0.0, "altitude", passed, failed);

    let slp = bmp_full.sea_level_pressure(0.0).unwrap_or(0.0);
    check_true!(slp >= 900.0 && slp <= 1100.0, "sea_level_pressure", passed, failed);

    println!("===DONE: {} passed, {} failed===", passed, failed);
    std::process::exit(if failed == 0 { 0 } else { 1 });
}

File: go/tests/pressure/bmp180_test/main.go
//go:build linux && !tinygo

// BMP180 hardware test — Linux host.
//
// Reads from /dev/i2c-N and runs the BMP180 check sequence. Prints
// PASS/FAIL per check and ends with the standard ===DONE: ... === line.
// Exits 0 on full pass, 1 on any failure.
package main

import (
	"fmt"
	"os"
	"strconv"

	"github.com/tuhde/Periph/go/periph/chips/pressure"
	"github.com/tuhde/Periph/go/periph/transport"
)

func main() {
	bus, err := strconv.Atoi(envOr("I2C_BUS", "1"))
	if err != nil {
		fmt.Fprintln(os.Stderr, "I2C_BUS:", err)
		os.Exit(2)
	}

	tr, err := transport.NewI2CTransport(bus, 0x77)
	if err != nil {
		fmt.Fprintln(os.Stderr, "transport:", err)
		os.Exit(2)
	}
	defer tr.Close()

	passed, failed := 0, 0
	check := func(label string, cond bool) {
		if cond {
			fmt.Println("PASS", label)
			passed++
		} else {
			fmt.Println("FAIL", label)
			failed++
		}
	}

	chip, err := pressure.NewBmp180Minimal(tr)
	if err != nil {
		fmt.Fprintln(os.Stderr, "new:", err)
		os.Exit(2)
	}

	t, err := chip.Temperature()
	check("temperature_range", err == nil && t >= -40.0 && t <= 85.0)

	p, err := chip.Pressure()
	check("pressure_range", err == nil && p >= 300.0 && p <= 1100.0)

	tr2, err := transport.NewI2CTransport(bus, 0x77)
	if err != nil {
		fmt.Fprintln(os.Stderr, "transport full:", err)
		os.Exit(2)
	}
	defer tr2.Close()

	full, err := pressure.NewBmp180Full(tr2)
	if err != nil {
		fmt.Fprintln(os.Stderr, "new full:", err)
		os.Exit(2)
	}
	check("default_oss", full.Oversampling() == pressure.OssUlp)
	full.SetOversampling(pressure.OssHighRes)
	check("set_oss", full.Oversampling() == pressure.OssHighRes)

	alt, err := full.AltitudeAt(1013.25)
	check("altitude_range", err == nil && alt >= -500.0 && alt <= 9000.0)

	slp, err := full.SeaLevelPressure(0.0)
	check("sea_level_pressure", err == nil && slp >= 900.0 && slp <= 1100.0)

	cid, err := full.ChipID()
	check("chip_id", err == nil && cid == 0x55)

	if err := full.Reset(); err != nil {
		fmt.Fprintln(os.Stderr, "reset:", err)
	}
	check("reset", err == nil)

	fmt.Printf("===DONE: %d passed, %d failed===\n", passed, failed)
	if failed != 0 {
		os.Exit(1)
	}
}

func envOr(k, def string) string {
	if v, ok := os.LookupEnv(k); ok {
		return v
	}
	return def
}
